src/errors/error_kind.rs

```diff
@@ -0,0 +1,319 @@
+use miette::Diagnostic;
+use thiserror::Error as ThisError;
+
+use crate::auth::AuthenticationScheme;
+use crate::StatusCode;
+
+/// Error kind for http-types
+#[derive(Debug, Diagnostic, ThisError)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum Error {
+    /// This only happens for APIs which support `TryInto` for an argument.
+    #[error("An argument failed to convert during TryInto: {}", .0)]
+    ArgTryIntoError(Box<dyn std::error::Error + Send + Sync + 'static>),
+
+    #[error("HTTP Method was unrecognized: {}", .0)]
+    HttpMethodUnrecognized(String),
+
+    #[error("I/O error: {}", .0)]
+    IO(#[from] std::io::Error),
+
+    #[error("Invalid Status Code: {}", .0)]
+    StatusCodeInvalid(u16),
+
+    #[error("Querystring deserialization error: {}", .0)]
+    QueryDeserialize(serde_qs::Error),
+    #[error("Querystring serialization error: {}", .0)]
+    QuerySerialize(serde_qs::Error),
+
+    #[error("Body error: {}", .0)]
+    Body(BodyError),
+
+    #[error("Header error: {}", .0)]
+    Header(HeaderError),
+
+    #[cfg(feature = "hyperium_http")]
+    #[error("URL Parse error: {}", .0)]
+    URLParse(#[from] url::ParseError),
+
+    #[cfg(feature = "hyperium_http")]
+    #[error("Hyperium HTTP error: {}", .0)]
+    HyperiumHttp(#[from] http::Error),
```

Comment:
Could we just bundle this into a case for third-party library errors, and support downcasting? That would avoid exposing third-party types in our API, which could introduce breakage if those libraries bump versions.

Reply:
@joshtriplett What would that look like? Specifically the "bundle this into a case for third-party library errors" part of that